source/extensions/transport_sockets/capture/capture.cc

```diff
@@ -0,0 +1,110 @@
+#include "extensions/transport_sockets/capture/capture.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/network/utility.h"
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Capture {
+
+CaptureSocket::CaptureSocket(const std::string& path_prefix, bool text_format,
+                             Network::TransportSocketPtr&& transport_socket)
+    : path_prefix_(path_prefix), text_format_(text_format),
+      transport_socket_(std::move(transport_socket)) {}
+
+void CaptureSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) {
+  callbacks_ = &callbacks;
+  transport_socket_->setTransportSocketCallbacks(callbacks);
+}
+
+std::string CaptureSocket::protocol() const { return transport_socket_->protocol(); }
+
+bool CaptureSocket::canFlushClose() { return transport_socket_->canFlushClose(); }
+
+void CaptureSocket::closeSocket(Network::ConnectionEvent event) {
+  // The caller should have invoked setTransportSocketCallbacks() prior to this.
+  ASSERT(callbacks_ != nullptr);
+  auto* connection = trace_.mutable_connection();
+  connection->set_id(callbacks_->connection().id());
+  Network::Utility::addressToProtobufAddress(*callbacks_->connection().localAddress(),
+                                             *connection->mutable_local_address());
+  Network::Utility::addressToProtobufAddress(*callbacks_->connection().remoteAddress(),
+                                             *connection->mutable_remote_address());
+  const std::string path = fmt::format("{}_{}.{}", path_prefix_, callbacks_->connection().id(),
+                                       text_format_ ? "pb_text" : "pb");
+  ENVOY_LOG_MISC(debug, "Writing socket trace for [C{}] to {}", callbacks_->connection().id(),
+                 path);
+  ENVOY_LOG_MISC(trace, "Socket trace for [C{}]: {}", callbacks_->connection().id(),
+                 trace_.DebugString());
+  std::ofstream proto_stream(path);
+  if (text_format_) {
+    proto_stream << trace_.DebugString();
+  } else {
+    trace_.SerializeToOstream(&proto_stream);
+  }
+  transport_socket_->closeSocket(event);
+}
+
+Network::IoResult CaptureSocket::doRead(Buffer::Instance& buffer) {
+  Network::IoResult result = transport_socket_->doRead(buffer);
+  if (result.bytes_processed_ > 0) {
+    // TODO(htuch): avoid linearizing
+    char* data = static_cast<char*>(buffer.linearize(buffer.length())) +
+                 (buffer.length() - result.bytes_processed_);
+    auto* event = trace_.add_events();
+    event->mutable_timestamp()->MergeFrom(Protobuf::util::TimeUtil::NanosecondsToTimestamp(
+        std::chrono::duration_cast<std::chrono::nanoseconds>(
+            std::chrono::system_clock::now().time_since_epoch())
+            .count()));
+    event->mutable_read()->set_data(data, result.bytes_processed_);
```

Comment:
Should we be capturing end_stream here? What about the action? Depending on what we're trying to replay, these may be needed.

Reply:
I'll capture half-close and add a TODO for the test; I think as long as we can grow the behavior and protos to capture these things in a backwards compatible way, it's fine to do this incrementally.